sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,526 @@
+      *var = csr_value;                    \
+    } else {                               \
+      csr_value = *var;                    \
+      PMP_CSR_WRITE(csr, csr_value);       \
+    }                                      \
+  } while (false)
+
+typedef enum pmp_csr_access_type {
+  kPmpCsrAccessTypeRead = 0,
+  kPmpCsrAccessTypeWrite,
+} pmp_csr_access_type_t;
+
+static const bitfield_field32_t kPmpCfgModeField = {
+    .mask = PMP_CFG_CSR_MODE_MASK, .index = PMP_CFG_CSR_A,
+};
+
+// TODO - X macros as suggested by @lenary.
+
+// TODO - description.
+PMP_WARN_UNUSED_RESULT
+static pmp_region_configure_result_t pmp_cfg_csr_rw(
+    pmp_region_index_t region, pmp_csr_access_type_t access,
+    uint32_t *variable_address) {
+  switch (region) {
+    case 0:
+    case 1:
+    case 2:
+    case 3:
+      PMP_CSR_RW(access, pmpcfg0, variable_address);
+      break;
+    case 4:
+    case 5:
+    case 6:
+    case 7:
+      PMP_CSR_RW(access, pmpcfg1, variable_address);
+      break;
+    case 8:
+    case 9:
+    case 10:
+    case 11:
+      PMP_CSR_RW(access, pmpcfg2, variable_address);
+      break;
+    case 12:
+    case 13:
+    case 14:
+    case 15:
+      PMP_CSR_RW(access, pmpcfg3, variable_address);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+// TODO - description.
+PMP_WARN_UNUSED_RESULT
+static bool pmp_addr_csr_rw(pmp_region_index_t region,
+                            pmp_csr_access_type_t access,
+                            uint32_t *variable_address) {
+  switch (region) {
+    case 0:
+      PMP_CSR_RW(access, pmpaddr0, variable_address);
+      break;
+    case 1:
+      PMP_CSR_RW(access, pmpaddr1, variable_address);
+      break;
+    case 2:
+      PMP_CSR_RW(access, pmpaddr2, variable_address);
+      break;
+    case 3:
+      PMP_CSR_RW(access, pmpaddr3, variable_address);
+      break;
+    case 4:
+      PMP_CSR_RW(access, pmpaddr4, variable_address);
+      break;
+    case 5:
+      PMP_CSR_RW(access, pmpaddr5, variable_address);
+      break;
+    case 6:
+      PMP_CSR_RW(access, pmpaddr6, variable_address);
+      break;
+    case 7:
+      PMP_CSR_RW(access, pmpaddr7, variable_address);
+      break;
+    case 8:
+      PMP_CSR_RW(access, pmpaddr8, variable_address);
+      break;
+    case 9:
+      PMP_CSR_RW(access, pmpaddr9, variable_address);
+      break;
+    case 10:
+      PMP_CSR_RW(access, pmpaddr10, variable_address);
+      break;
+    case 11:
+      PMP_CSR_RW(access, pmpaddr11, variable_address);
+      break;
+    case 12:
+      PMP_CSR_RW(access, pmpaddr12, variable_address);
+      break;
+    case 13:
+      PMP_CSR_RW(access, pmpaddr13, variable_address);
+      break;
+    case 14:
+      PMP_CSR_RW(access, pmpaddr14, variable_address);
+      break;
+    case 15:
+      PMP_CSR_RW(access, pmpaddr15, variable_address);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+// TODO - description.
```

Comment:
I assume you're not merging with these TODOs? Can you please resolve these before merging?

Reply:
Done.